src/lib/blockstore.js

```diff
@@ -109,69 +108,62 @@ export class BatchingR2Blockstore extends R2Blockstore {
 
   async #processBatch () {
     console.log('processing batch')
-    const batches = this.#batches
-    const batchBlocks = this.#batchBlocks
-    this.#batches = new Map()
-    this.#batchBlocks = new Map()
+    const batcher = this.#batcher
+    this.#batcher = new OrderedCarBlockBatcher()
+    const pendingBlocks = this.#pendingBlocks
+    this.#pendingBlocks = new Map()
+
+    while (true) {
+      const batch = batcher.next()
+      if (!batch.length) break
+
+      batch.sort((a, b) => a.offset - b.offset)
+
+      const { carCid } = batch[0]
+      const carPath = `${carCid}/${carCid}.car`
+      const range = {
+        offset: batch[0].offset,
+        length: batch[batch.length - 1].offset - batch[0].offset + MAX_ENCODED_BLOCK_LENGTH
+      }
 
-    for (const [carCid, batcher] of batches) {
-      console.log(`processing batch for ${carCid}`)
-      while (true) {
-        const batch = batcher.next()
-        if (!batch) break
-        const carPath = `${carCid}/${carCid}.car`
-        const range = { offset: batch[0], length: batch[batch.length - 1] - batch[0] + MAX_BLOCK_LENGTH }
-
-        console.log(`waiting allocation for ${range.length} bytes...`)
-        await this._memoryBudget.request(range.length)
-
-        console.log(`fetching ${batch.length} blocks from ${carCid} (${range.length} bytes @ ${range.offset})`)
-        const res = await this._dataBucket.get(carPath, { range })
-        if (!res) {
-          for (const blocks of batchBlocks.values()) {
-            blocks.forEach(b => b.resolve())
-          }
-          return
+      console.log(`fetching ${batch.length} blocks from ${carCid} (${range.length} bytes @ ${range.offset})`)
+      const res = await this._dataBucket.get(carPath, { range })
+      if (!res) {
+        for (const blocks of pendingBlocks.values()) {
+          blocks.forEach(b => b.resolve())
         }
+        return
```

Comment:
plz comment to explain this! From convo this is to furfil the contract of returning undefined if we dont have the block